### PR TITLE
updates to support typed errors

### PR DESCRIPTION
**Pull request by ettec**

### CODE REVIEW

File: packages/cre-sdk/src/capabilities/errors/error.ts [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+		private readonly errorCode: ErrorCode,
+	) {
+		super(
+			errorCode === UnrecognisedErrorCode
+				? detail
+				: `[${errorCode}]${errorCodeToStringValue(errorCode)}: ${detail}`,
+		)
+	}
+
+	visibility(): Visibility {
```
Comment (russell-stern) [inline packages/cre-sdk/src/capabilities/errors/error.ts]: These should be readonly properties instead of functions. You can look at the CapabilityError class in `capabaility-error.ts` for reference

File: packages/cre-sdk/src/capabilities/errors/error.ts [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+		case 'Public':
+			return VisibilityPublic
+		case 'Private':
+			return VisibilityPrivate
+		default:
+			return -1
+	}
+}
+
+export interface ICreCapabilityError {
```
Comment (russell-stern) [inline packages/cre-sdk/src/capabilities/errors/error.ts]: You can drop the interface and just extend Error

File: packages/cre-sdk/src/capabilities/errors/error.ts [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+	override toString(): string {
+		return this.message
+	}
+}
+
+export function isCreCapabilityError(err: unknown): err is CreCapabilityError {
+	return err instanceof CreCapabilityError
+}
+
+export function newCreCapabilityError(
```
Comment (russell-stern) [inline packages/cre-sdk/src/capabilities/errors/error.ts]: Typescript doesn't generally use these Factories, you can use `new CreCapabilityError(...)` and remove this function

File: packages/cre-sdk/src/capabilities/errors/error.ts (hunk 1 above)
Comment (russell-stern) [reply packages/cre-sdk/src/capabilities/errors/error.ts]:
```
export class CreCapabilityError extends Error {
  readonly name = 'CreCapabilityError'
  constructor(
    readonly detail: string,
    readonly visibility: Visibility,
    readonly origin: Origin,
    readonly code: ErrorCode,
  ) {
    super(
      code === UnrecognisedErrorCode
        ? detail
        : `[${code}]${formatErrorCode(code)}: ${detail}`,
    )
  }
}
```

File: packages/cre-sdk/src/sdk/impl/runtime-impl.test.ts [hunk 4]
```diff
@@ -219,16 +219,35 @@ describe('test runtime', () => {
...
+			})
+
+			const runtime = new RuntimeImpl<unknown>({}, 1, helpers, anyMaxSize)
+			const workflowAction1 = new BasicActionCapability()
+			const call1 = workflowAction1.performAction(
+				runtime,
+				create(InputsSchema, { inputThing: true }),
 			)
+
+			expect(() => call1.result()).toThrow('[4]DeadlineExceeded: capability failed')
```
Comment (russell-stern) [inline packages/cre-sdk/src/sdk/impl/runtime-impl.test.ts]: This is only checking if the string matches. We should check if the error is actually a CreCapabilityError:
`expect(() => call1.result()).toThrow(CreCapabilityError)`

File: packages/cre-sdk/src/standard_tests/capability_errors/test.ts [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+import { BasicActionCapability } from '@cre/generated-sdk/capabilities/internal/basicaction/v1/basicaction_sdk_gen'
+import { BasicCapability as BasicTriggerCapability } from '@cre/generated-sdk/capabilities/internal/basictrigger/v1/basic_sdk_gen'
+import { cre, type Runtime } from '@cre/sdk/cre'
+import { Runner } from '@cre/sdk/wasm'
+
+const checkCapabilityErrors = (runtime: Runtime<Uint8Array>) => {
+	const basicAction = new BasicActionCapability()
+	const input = { inputThing: true }
+
+	while (true) {
```
Comment (russell-stern) [inline packages/cre-sdk/src/standard_tests/capability_errors/test.ts:11]: This isn't enumerating over all errors it's just checking if it's a CreCapabilityError or an UnrecognisedErrorCode. Can we export all the errors and run through them here like we're doing in the Go sdk?

**Review by russell-stern — CHANGES_REQUESTED** [review]

File: packages/cre-sdk/src/capabilities/errors/error-serialization.ts [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+import { fromErrorCodeString } from './error-codes'
+
+const errorMessageSeparator = ':'
+
+/**
+ * Parses errorMsg in the capability error wire format.
+ * If errorMsg is not a serialized capability error, a plain Error is returned.
+ */
+export function deserializeErrorFromString(errorMsg: string): Error {
+	const parts = errorMsg.split(errorMessageSeparator, 4)
```
Comment (ernest-nowacki) [inline packages/cre-sdk/src/capabilities/errors/error-serialization.ts]: Here I believe we have an edge case (potentially) - I'm not sure how exactly the error details (messages) would look like but if any of them includes colon it would basically break our split. 

So imagine that actual error is `host: connection lost`. 

Our end result would lost `connection lost` information, error details would only have `host` in it.

It looks like the format is 3 fields + the reminding part is an actual error message. One simple option that I could think about is just to rejoin everything when building detail field. Will add a snippet next to that line.

We could also add a unit test for this edge case.

File: packages/cre-sdk/src/capabilities/errors/error-serialization.ts [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+
+	if (parts.length < 4) {
+		return new Error(errorMsg)
+	}
+
+	const visibility = fromVisibilityString(parts[0])
+	const origin = fromOriginString(parts[1])
+	const errorCode = fromErrorCodeString(parts[2])
+	const detail = parts[3]
```
Comment (ernest-nowacki) [inline packages/cre-sdk/src/capabilities/errors/error-serialization.ts]:
```suggestion
	const detail = parts.slice(3).join(errorMessageSeparator)
```

File: packages/cre-sdk/Makefile [hunk 8]
```diff
@@ -1,4 +1,4 @@
-COMMON_VERSION ?= cre-std-tests@0.6.0
+COMMON_VERSION ?= c5ae21941dc8122572fa2702afc5f9b23b2ce762
```
Comment (ernest-nowacki) [inline packages/cre-sdk/Makefile]: Is this intentional to be kept as a commit hash rather then some tag? I understand we wanted to bump but I thought maybe just to latest tagged version?

Comment (ettec) [reply packages/cre-sdk/Makefile]: I'll be updating to a tag before merge

File: packages/cre-sdk/src/standard_tests/capability_errors/test.ts (hunk 5 above)
Comment (ettec) [reply packages/cre-sdk/src/standard_tests/capability_errors/test.ts:11]: this test is passed all error codes from common as part of the standard test (until it gets "Done").  If the sdk is missing any we will get UnrecogniedErrorCode which tells us the SDK is out of sync and test will fail, thats the purpose of it.  This is the same as in the go sdk: standard_tests/capability_errors/main_wasip1.go

File: packages/cre-sdk/src/capabilities/errors/error-serialization.ts (hunk 6 above)
Comment (ettec) [reply packages/cre-sdk/src/capabilities/errors/error-serialization.ts]: In the case of this message:

host: connection lost.

It would parse to less than 4 parts so it would return the full message (line 13)

Also, barring some bug in the system the only stringified error we should get back in the capability wrappers should be capability errors which will always have the 'Visibility:Origin:ErrorCode:'  prepend, so any colons after that will be fine.   There is a possibly a bit of an edge case where for some reason we get an error string  in the capability wrappers that is not a serialized capability error that contains 3+ colons - IMO its unlikely.

File: packages/cre-sdk/src/capabilities/errors/error.ts (hunk 3 above)
Comment (ettec) [reply packages/cre-sdk/src/capabilities/errors/error.ts]: sorted

File: packages/cre-sdk/src/capabilities/errors/error.ts (hunk 2 above)
Comment (ettec) [reply packages/cre-sdk/src/capabilities/errors/error.ts]: done

File: packages/cre-sdk/src/capabilities/errors/error.ts (hunk 1 above)
Comment (ettec) [reply packages/cre-sdk/src/capabilities/errors/error.ts]: done

File: packages/cre-sdk/src/capabilities/errors/error-serialization.ts (hunk 7 above)
Comment (ettec) [reply packages/cre-sdk/src/capabilities/errors/error-serialization.ts]: done, nice catch, a subtle difference between go and typescripts split method with parts limit.  Also added a unit test to check colons are preserved.

File: packages/cre-sdk/src/sdk/impl/runtime-impl.test.ts (hunk 4 above)
Comment (ettec) [reply packages/cre-sdk/src/sdk/impl/runtime-impl.test.ts]: done

File: packages/cre-sdk/src/sdk/utils/capabilities/capability-error.ts [hunk 9]
```diff
@@ -13,7 +13,7 @@ export class CapabilityError extends Error {
 		},
 	) {
 		super(message)
-		this.name = 'CapabilityError'
+		this.name = 'CapabilityRuntimeError'
```
Comment (russell-stern) [inline packages/cre-sdk/src/sdk/utils/capabilities/capability-error.ts]: The name will have to stay the same to be backwards compatible. Instead of renaming the class we can create the alias in the other direction
```
export class CapabilityError extends Error { ... }
export { CapabilityError as CapabilityRuntimeError }
```

**Review by russell-stern — CHANGES_REQUESTED** [review]

**Comment by ernest-nowacki** [comment]
There's also some lint errors related to formatting - those can probably be automatically fixed by running `bun format` from monorepo root.

File: packages/cre-sdk/src/sdk/utils/capabilities/capability-error.ts (hunk 9 above)
Comment (ettec) [reply packages/cre-sdk/src/sdk/utils/capabilities/capability-error.ts]: done

File: packages/cre-sdk/Makefile (hunk 8 above)
Comment (ettec) [reply packages/cre-sdk/Makefile]: updated

**Review by ernest-nowacki — APPROVED** [review]

**Review by russell-stern — APPROVED** [review]
